File: internal/request/request.go
package request

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"strconv"

	"github.com/taham8875/http-from-tcp/internal/headers"
)

const (
	StateInit = iota
	StateParsingHeaders
	StateParsingBody
	StateDone
)

type Request struct {
	RequestLine RequestLine
	Headers     headers.Headers
	Body        string
	State       int
}

type RequestLine struct {
	HttpVersion   string
	RequestTarget string
	Method        string
}

const bufferSize = 1024
const SEPARATOR = "\r\n"

var ERROR_MALFORMED_REQUEST = fmt.Errorf("malformed request")
var ERROR_UNSUPPORTED_HTTP_VERSION = fmt.Errorf("unsupported http version")
var ERROR_PARSE_WHEN_DONE = fmt.Errorf("trying to parse when already done")
var ERROR_INVALID_STATE = fmt.Errorf("invalid state")

func newRequest() *Request {
	return &Request{
		State: StateInit,
	}
}

func RequestFromReader(reader io.Reader) (*Request, error) {
	buf := make([]byte, bufferSize)
	readToIndex := 0

	request := newRequest()

	for request.State != StateDone {
		if readToIndex >= len(buf) {
			newBuf := make([]byte, len(buf)*2)
			copy(newBuf, buf)
			buf = newBuf
		}

		n, err := reader.Read(buf[readToIndex:])

		if err != nil {
			if err == io.EOF {
				// if we are expecting a body, ensure we got it all
				if request.State == StateParsingBody {
					if v, ok := request.Headers.Get("content-length"); ok && len(v) > 0 {
						if cl, err := strconv.Atoi(v); err == nil && cl >= 0 {
							if len(request.Body) < cl {
								return nil, ERROR_MALFORMED_REQUEST
							}
						} else {
							return nil, ERROR_MALFORMED_REQUEST
						}
					}
				}
				request.State = StateDone
				break
			}

			return nil, err
		}

		readToIndex += n

		bytesConsumed, err := request.parse(buf[:readToIndex])

		if err != nil {
			return nil, err
		}

		if bytesConsumed > 0 {
			copy(buf, buf[bytesConsumed:])
			readToIndex -= bytesConsumed
		}
	}

	return request, nil
}

func (r *Request) parse(data []byte) (int, error) {
	totalBytesParsed := 0
	for r.State != StateDone {
		n, err := r.parseSingle(data[totalBytesParsed:])
		if err != nil {
			return totalBytesParsed, err
		}

		if n == 0 {
			// need more data
			return totalBytesParsed, nil
		}

		totalBytesParsed += n

	}

	return totalBytesParsed, nil
}

func (r *Request) parseSingle(data []byte) (int, error) {
	switch r.State {
	case StateInit:
		requestLine, bytesConsumed, err := parseRequestLines(data)
		if err != nil {
			return 0, err
		}

		if bytesConsumed == 0 {
			return 0, nil
		}

		r.RequestLine = *requestLine
		r.State = StateParsingHeaders

		return bytesConsumed, nil
	case StateParsingHeaders:
		if r.Headers == nil {
			r.Headers = headers.NewHeaders()
		}

		n, done, err := r.Headers.Parse(data)
		if err != nil {
			return 0, err
		}

		if n == 0 {
			// need more date
			return 0, nil
		}

		if done {
			r.State = StateParsingBody
		}

		return n, nil

	case StateParsingBody:
		contentLengthValue, ok := r.Headers.Get("content-length")

		// if no content-length, assume no body and finish
		if !ok || len(contentLengthValue) == 0 || contentLengthValue == "0" {
			r.State = StateDone
			return 0, nil
		}

		// convert contentLength to int from contentLengthValue
		contentLength, err := strconv.Atoi(contentLengthValue)
		if err != nil || contentLength < 0 {
			return 0, ERROR_MALFORMED_REQUEST
		}

		r.Body += string(data)

		if len(r.Body) > contentLength {
			return len(data), ERROR_MALFORMED_REQUEST
		}

		if len(r.Body) == contentLength {
			r.State = StateDone
			return len(data), nil
		}

		return len(data), nil

	case StateDone:
		return 0, ERROR_PARSE_WHEN_DONE
	default:
		return 0, errors.New("invalid state")
	}
}

func parseRequestLines(requestBytes []byte) (*RequestLine, int, error) {
	rnIdx := bytes.Index(requestBytes, []byte(SEPARATOR))
	if rnIdx == -1 {
		// need more data for a full request line
		return nil, 0, nil
	}

	line := requestBytes[:rnIdx]
	parts := bytes.Split(line, []byte(" "))
	if len(parts) != 3 {
		return nil, 0, ERROR_MALFORMED_REQUEST
	}

	httpVersion := parts[2]
	if string(httpVersion) != "HTTP/1.1" {
		return nil, 0, ERROR_UNSUPPORTED_HTTP_VERSION
	}

	httpVersion = bytes.TrimPrefix(httpVersion, []byte("HTTP/"))

	return &RequestLine{
		Method:        string(parts[0]),
		RequestTarget: string(parts[1]),
		HttpVersion:   string(httpVersion),
	}, rnIdx + len(SEPARATOR), nil
}

File: python-version/internal/request/request.py
"""
HTTP Request Parser — parseia requests HTTP/1.1 a partir de um stream TCP raw.

Usa uma máquina de estados:
  INIT -> PARSING_HEADERS -> PARSING_BODY -> DONE
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum, auto
from typing import BinaryIO

from internal.headers import Headers, InvalidHeaderError

CRLF = b"\r\n"
BUFFER_SIZE = 1024


class ParseState(IntEnum):
    INIT = auto()
    PARSING_HEADERS = auto()
    PARSING_BODY = auto()
    DONE = auto()


class MalformedRequestError(Exception):
    pass


class UnsupportedHTTPVersionError(Exception):
    pass


@dataclass
class RequestLine:
    method: str = ""
    request_target: str = ""
    http_version: str = ""


@dataclass
class Request:
    request_line: RequestLine = field(default_factory=RequestLine)
    headers: Headers = field(default_factory=Headers)
    body: bytes = b""
    state: ParseState = ParseState.INIT


def request_from_reader(reader: BinaryIO) -> Request:
    """
    Lê um request HTTP completo de um reader (socket, BytesIO, etc).
    Parseia incrementalmente à medida que dados chegam.

    Usa read1() quando disponível (BufferedReader de sockets) para não
    bloquear esperando o buffer inteiro — equivale ao behavior de Read() em Go.
    """
    buf = bytearray(BUFFER_SIZE)
    read_to_index = 0
    req = Request()

    # read1 retorna assim que QUALQUER dado está disponível (não bloqueia pro buffer cheio)
    _read = getattr(reader, "read1", None) or reader.read

    while req.state != ParseState.DONE:
        if read_to_index >= len(buf):
            buf.extend(bytearray(len(buf)))

        chunk = _read(BUFFER_SIZE)

        if not chunk:
            if req.state == ParseState.PARSING_BODY:
                cl_str, has_cl = req.headers.get_header("content-length")
                if has_cl and cl_str:
                    cl = int(cl_str)
                    if len(req.body) < cl:
                        raise MalformedRequestError(
                            "body shorter than content-length"
                        )
            req.state = ParseState.DONE
            break

        end = read_to_index + len(chunk)
        if end > len(buf):
            buf.extend(bytearray(end - len(buf)))
        buf[read_to_index:end] = chunk
        read_to_index = end

        bytes_consumed = _parse(req, bytes(buf[:read_to_index]))

        if bytes_consumed > 0:
            buf[:read_to_index - bytes_consumed] = buf[bytes_consumed:read_to_index]
            read_to_index -= bytes_consumed

    return req


def _parse(req: Request, data: bytes) -> int:
    total = 0
    while req.state != ParseState.DONE:
        n = _parse_single(req, data[total:])
        if n == 0:
            break
        total += n
    return total


def _parse_single(req: Request, data: bytes) -> int:
    if req.state == ParseState.INIT:
        return _parse_request_line(req, data)
    elif req.state == ParseState.PARSING_HEADERS:
        return _parse_headers(req, data)
    elif req.state == ParseState.PARSING_BODY:
        return _parse_body(req, data)
    elif req.state == ParseState.DONE:
        raise MalformedRequestError("trying to parse when already done")
    else:
        raise MalformedRequestError("invalid state")


def _parse_request_line(req: Request, data: bytes) -> int:
    rn_idx = data.find(CRLF)
    if rn_idx == -1:
        return 0

    line = data[:rn_idx]
    parts = line.split(b" ")
    if len(parts) != 3:
        raise MalformedRequestError("request line must have 3 parts")

    method, target, version = parts

    if version != b"HTTP/1.1":
        raise UnsupportedHTTPVersionError(f"unsupported: {version.decode()}")

    req.request_line = RequestLine(
        method=method.decode("ascii"),
        request_target=target.decode("ascii"),
        http_version="1.1",
    )
    req.state = ParseState.PARSING_HEADERS
    return rn_idx + 2


def _parse_headers(req: Request, data: bytes) -> int:
    n, done = req.headers.parse(data)
    if n == 0:
        return 0
    if done:
        req.state = ParseState.PARSING_BODY
    return n


def _parse_body(req: Request, data: bytes) -> int:
    cl_str, has_cl = req.headers.get_header("content-length")

    if not has_cl or not cl_str or cl_str == "0":
        req.state = ParseState.DONE
        return 0

    try:
        content_length = int(cl_str)
    except ValueError:
        raise MalformedRequestError("invalid content-length")

    if content_length < 0:
        raise MalformedRequestError("negative content-length")

    req.body += data

    if len(req.body) > content_length:
        raise MalformedRequestError("body exceeds content-length")

    if len(req.body) == content_length:
        req.state = ParseState.DONE

    return len(data)
